osu.Game.Rulesets.Taiko/Skinning/Legacy/LegacyCirclePiece.cs

```diff
@@ -96,5 +126,15 @@ private void updateAccentColour()
         {
             backgroundLayer.Colour = LegacyColourCompatibility.DisallowZeroAlpha(accentColour);
         }
+
+        protected override void OnNewBeat(int beatIndex, TimingControlPoint timingPoint, EffectControlPoint effectPoint, ChannelAmplitudes amplitudes)
+        {
+            if (!enableAnimations)
+                return;
+
+            animationFrame = beatIndex % 4 == 0 || (beatIndex - 1) % 4 == 0 ? 1 : 0;
```

Comment:
Do you mind explaining further what this is supposed to be doing? I can't see anything of this sort in stable source.

Reply:
This part determines which frame of the animation to show on every `OnNewBeat` call. Because it is called twice as much as usual (`Divisor` is doubled, see PR description for reason), `animationFrame`'s result needs to be the same for two calls in a row. Below image might make this more clear, each x is an `OnNewBeat` call:

![Screenshot 2022-11-12 160358](https://user-images.githubusercontent.com/56885706/201480339-71612088-d5f6-42ab-b845-0d643c14815a.jpg)

If I didn't need to double `Divisor`, I would be able to do this which might be more in line with stable:
```
animationFrame = beatIndex % 2 == 0 ? 1 : 0;
```

Reply:
That simplified variant looks closer to what I'd expect to see.

I could have a closer look at this but I'd like these workarounds to not exist if possible, or to be properly signposted if they need to exist. Are you sure that stable doesn't have the same "animation tackiness"?

Reply:
To help understand what I meant by "tackiness", notice the animations at the very right edge of the playfield start too late without the `Divisor` workaround applied (this does not happen in stable):
https://user-images.githubusercontent.com/56885706/201521627-5002ff88-b236-457a-bee5-ab769a8b0e8b.mp4

I'm not sure how I would go about solving this, for this to work without a workaround I think the hitobjects would need to be created earlier so they have a bigger margin of time to be affected by `OnNewBeat` before they're visible to the player. I'm not sure if this is possible/wanted though. What are your thoughts on this issue?

Reply:
To confirm, this animation discrepancy is what you're trying to fix, correct?

![image](https://user-images.githubusercontent.com/20418176/202872800-5c8cbc56-a24f-48bb-a5d9-d34533d7196b.png)

If so, then it may be a decent idea to ditch `BeatSyncedContainer` and kinda copy stable's implementation a little, which is:

```csharp
            bool doCircleAnimation = ComboCounter.HitCombo >= 50;


            if (doCircleAnimation)
            {
                circlesFrame = (ComboCounter.HitCombo >= 150 ? MouthOpen(2) : MouthOpen(1)) ? 0 : 1;


                hitObjectManager.hitObjectsMinimal.ForEach(ho =>
                {
                    SliderTaiko s = ho as SliderTaiko;
                    if (s != null)
                        s.sliderStartCircle.SpriteHitCircle2.CurrentFrame = circlesFrame;
                    else
                    {
                        HitCircleTaiko h = ho as HitCircleTaiko;
                        if (h != null)
                            h.SpriteHitCircle2.CurrentFrame = circlesFrame;
                    }
                });
            }
```

wherein

```csharp
        internal static bool MouthOpen(int multiplier)
        {
            double beatLength = AudioEngine.BeatLength;
            return AudioEngine.SyncTime % ((beatLength * 2) / multiplier)
                   >= beatLength / multiplier;
        }
```

Note that all of the above is using very basic quantities, i.e. current time and beat length.

Maybe there's a way to move this 1:1 from stable using bits and pieces alreadyh available in lazer. Tough to say.

Reply:
Yes, that's what I meant. I was able to port the stable implementation over to lazer without much effort (thanks for giving me the code) and it seems to work nicely. I still needed `LegacyCirclePiece` to be a `BeatSyncedContainer` so that I can easily get `BeatLength`, but apart from that `OnNewBeat` and the `Divisor` workaround have been removed! Let me know if below looks okay since it basically changes this entire PR, ready to push.

``` csharp
protected override void Update()
        {
            base.Update();

            if (currentCombo.Value >= 150)
            {
                enableAnimations = true;
                multiplier = 2;
            }
            else if (currentCombo.Value >= 50)
            {
                enableAnimations = true;
                multiplier = 1;
            }
            else
            {
                enableAnimations = false;
                (foregroundLayer as IFramedAnimation)?.GotoFrame(0);
                return;
            }

            if (BeatSyncSource.ControlPoints != null)
                beatLength = BeatSyncSource.ControlPoints.TimingPointAt(LifetimeStart).BeatLength;

            animationFrame = Time.Current % ((beatLength * 2) / multiplier) >= beatLength / multiplier ? 0 : 1;

            if (enableAnimations)
                (foregroundLayer as IFramedAnimation)?.GotoFrame(animationFrame);
        }
```

Reply:
I believe you don't really need `BeatSyncedContainer` for `BeatLength` access, you could just as well resolve `IBeatSyncProvider` directly and thus cut the middleman.

I'd say that doing so and going with the alternate implementation above may be warranted in this particular case, since this is an unorthodox use case, and similar things [have been done in other places already](https://github.com/ppy/osu/blob/master/osu.Game/Storyboards/Drawables/DrawableStoryboardAnimation.cs#L131).